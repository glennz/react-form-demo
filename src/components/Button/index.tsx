import React from 'react';
import PropTypes from 'prop-types';

export const Button = (props: any)=> {
    const { id, type, text, className, click } = props;
    
    return (
        <button id={id}
            type={type}
            className={className}
            onClick={click}>
            {text}
        </button>
    );
};

Button.defaultProps = {
    type: "button",
};

Button.propTypes = {
    id: PropTypes.string.isRequired,
    type: PropTypes.string,
    className: PropTypes.string,
    text: PropTypes.string.isRequired,
    click: PropTypes.func,
    children: PropTypes.node
};