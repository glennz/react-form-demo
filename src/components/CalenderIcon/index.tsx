import React, { useState } from 'react';
import PropTypes from 'prop-types';
import icon from './images/calender.svg';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";

export const CalenderIcon = (props: any) => {
    const [datePicked, setDate] = useState(new Date());
    const { selectedDate, setDateValue } = props;

    if (!datePicked && selectedDate) {
        const date = new Date(selectedDate);
        setDate(date);     
    }

    const updateDate = (date: Date) => {
        setDate(date);
        setDateValue && setDateValue(date);
    };

    const ImageCustomInput = React.forwardRef((props: any, ref: any) => {
        return (
            <img src={icon} onClick={props.onClick} className="calender-icon" alt="" />
        )
    });
    
    return (        
        <DatePicker
        selected={datePicked}
        onChange={updateDate}
        customInput={<ImageCustomInput />}
        />
    );
};

CalenderIcon.propTypes = {
    selectedDate: PropTypes.any,
    setDateValue: PropTypes.func
};