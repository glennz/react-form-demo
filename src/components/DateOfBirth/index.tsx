import React from 'react';
import moment from 'moment';
import DateOfBirthPropsType from '../../shared/DateOfBirthPropsType';
import DateOfBirthType from '../../shared/DateOfBirthType';
import PropTypes from 'prop-types';
import utility from '../../constant/ulitily';

export const parseFields = (value: string, name: string) => {
  if (name === 'day') {
    return value.replace(/[^0-9]/g, '');
  }
  if (name === 'month') {
    return value.replace(/[^0-9a-zA-Z]/g, '');
  }
  if (name === 'year') {
    return value.replace(/[^0-9]/g, '');
  }
  if (name === 'parsed') {
    return value && moment.utc(value, 'YYYY-MM-DD').format();
  }
  return value;
};

export const formatFields = (value: string, name: string) => {
  let newVal = value;
  if (name === 'parsed' && newVal) {
    newVal = moment.utc(newVal).format('YYYY-MM-DD');
  }
  return newVal || '';
};

/* eslint-disable complexity */
export const DateOfBirth = (props: DateOfBirthPropsType) => {
  const {
    id, 
    name,   
    label,  
    touched,
    error,
    dateOfBirth,
    dateFormat,
    setTouched, 
    setValue,
  } = props;

  let value = dateOfBirth || {} as DateOfBirthType;

  const dateAuFormat = 'au';

  const inputProps = (fieldName: string) => ({
    className: 'dob-field',
    id: `${id}.${fieldName}`,
    name: `${name}.${fieldName}`,
    onBlur: () => {
        setTouched && setTouched(`${name}.parsed`, true);
    },
    onChange: (e: any) => {
      // const inputValue = parseFields(e.target.value, `${fieldName}`);
      // // setValue && setValue(`${name}.${fieldName}`, inputValue);
      const targetValue = e.target.value;
      switch(fieldName) {
          case 'day':
              value.day = utility.udpateControlData(targetValue);
              break;
          case 'month':
              value.month = utility.udpateControlData(targetValue);
              break;
          case 'year':
                value.year = utility.udpateControlData(targetValue);
                break;
          default:
              value.day = utility.initControlData;
              value.month = utility.initControlData;
              value.year = utility.initControlData;
              break;
      }

      const date = {
        day: value.day,
        month: value.month,
        year: value.year
      } as DateOfBirthType;

      setValue && setValue(date);
    }
  });

  const dayControl = () => {
    return ( 
      <input {...inputProps('day')} placeholder="dd" type="text" value={dateOfBirth.day.value} maxLength={2} />
    );
  };

  const monthControl = () => {
    return ( 
      <input {...inputProps('month')} placeholder="mm" type="text" value={dateOfBirth.month.value} maxLength={9} />
    );
  };

  const yearControl = () => {
    return ( 
      <input {...inputProps('year')} placeholder="yyyy" type="text" value={dateOfBirth.year.value} maxLength={4} />
    );
  };

  const auDateControl = () => {
    return (
      <div className="dob-inputs">
        {dayControl()} 
        <span className="splitter">/</span>
        {monthControl()} 
        <span className="splitter">/</span>
        {yearControl()}
      </div>
    );
  };

  const usaDateControl = () => {
    return (
      <div className="dob-inputs">
        {monthControl()}
        <span className="splitter">/</span>
        {dayControl()} 
        <span className="splitter">/</span>    
        {yearControl()}
      </div>
    );
  };

  const dateControl = () => {
    return (dateFormat === dateAuFormat) ? auDateControl() : usaDateControl();
  };

  const labelControl = () => {
    return (label) ? (<label>{label}</label>) : null;
  };

  return (
    <div className={`dob-inputs-section${touched && error ? ' error' : ''}`}>
      {labelControl()}
      {dateControl()}
      {touched && (error && <div className="error message">{error}</div>)}
    </div>
  );
};

DateOfBirth.defaultProps = {
  dateFormat: 'usa'
};

DateOfBirth.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,  
  label: PropTypes.string,
  dateOfBirth: PropTypes.any.isRequired,
  error: PropTypes.string,
  touched: PropTypes.bool,
  children: PropTypes.node,
  setTouched: PropTypes.func,
  setValue: PropTypes.func,
};