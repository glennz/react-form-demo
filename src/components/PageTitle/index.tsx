import React from 'react';
import PropTypes from 'prop-types';
import IPageTitleProps from './IPageTitleProps';

export const PageTitle = (props: IPageTitleProps) => {
  const { step, title, className } = props;

  return (
    <div className={className}>
        <span>{step}</span>
        <span>{title}</span>
    </div>
  );
};

PageTitle.propTypes = {
  step: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
};