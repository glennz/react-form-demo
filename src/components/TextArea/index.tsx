import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

export const TextArea = (props: any) => {
  const { id, label, name, className, value, error, onBlur, onChange, touched, rows, cols } = props;
  const classes = classnames({
    error: touched && error,
    'text-input': true
  });

  return (
    <div className={classes}>
      <label htmlFor={name}>{label}</label>
      <textarea
        className={`input ${className || ''}`}
        id={id}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        placeholder={label}
        value={value}
        rows={rows}
        cols={cols}
      />
      {touched && (error && <span className="message">{error}</span>)}
    </div>
  );
};

TextArea.defaultProps = {
  touched: false,
  error: ''
};

TextArea.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  className: PropTypes.string,
  error: PropTypes.string,  
  touched: PropTypes.bool,
  children: PropTypes.node,
  rows: PropTypes.string,
  cols: PropTypes.string,
};