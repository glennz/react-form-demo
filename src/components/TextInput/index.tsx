import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

export const TextInput = (props: any) => {
  const { id, type, label, name, className, value, error, onBlur, onChange, touched } = props;
  const classes = classnames({
    'text-input': true,
    error: touched && error    
  });

  return (
    <div className={classes}>
      <label htmlFor={name}>{label}</label>
      <input
        className={`input ${className || ''}`}
        id={id}
        type={type}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        placeholder={label}
        value={value}
      />
      {touched && (error && <div className="error-message">{error}</div>)}
    </div>
  );
};

TextInput.defaultProps = {
  type: "text",
  touched: false,
  error: ''
};

TextInput.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,  
  type: PropTypes.string,
  value: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  className: PropTypes.string,
  customChange: PropTypes.func,
  error: PropTypes.string,
  touched: PropTypes.bool,
  children: PropTypes.node
};