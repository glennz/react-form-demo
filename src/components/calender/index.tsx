import React from 'react';
import canlenderIcon from './images/canlender.svg';

export const Calender = (props: any)=> {
    return (
        <img src={canlenderIcon} alt="" />
    );
};
