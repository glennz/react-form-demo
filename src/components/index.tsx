import { PageTitle } from './PageTitle';
import { TextInput } from './TextInput';
import { TextArea } from './TextArea';
import { Button } from './Button';
import { DateOfBirth } from './DateOfBirth';

export {
    PageTitle,
    TextInput,
    TextArea,
    Button,
    DateOfBirth,
};