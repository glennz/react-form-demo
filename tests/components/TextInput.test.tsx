import React from 'react';
import { mount } from 'enzyme';

import { TextInput } from '../../src/components/TextInput';

describe('<TextInput />', () => {
  const fieldProps = {
    className: 'test',
    handleBlur: () => {},
    handleChange: () => {},
    id: 'test',
    label: 'Test label',
    name: 'test'
  };

  it('should render input field', () => {
    const textInput = mount(<TextInput {...fieldProps} />);
    expect(textInput.find('input')).toHaveLength(1);
    expect(textInput).toMatchSnapshot();
  });

  it('should render input field without error', () => {
    const touched = true;
    const textInput = mount(<TextInput {...fieldProps} touched={touched} className="" />);
    expect(textInput.find('input')).toHaveLength(1);
    expect(textInput.find('div')).toHaveLength(1);
    expect(textInput.find('.error-message')).toHaveLength(0);
    expect(textInput).toMatchSnapshot();
  });

  it('should render input field with error', () => {
    const touched = true;
    const error = 'error';
    const textInput = mount(<TextInput {...fieldProps} error={error} touched={touched} />);
    expect(textInput.find('input')).toHaveLength(1);
    expect(textInput.find('div')).toHaveLength(2);
    expect(textInput.find('.error-message')).toHaveLength(1);

    expect(textInput).toMatchSnapshot();
  });
});
