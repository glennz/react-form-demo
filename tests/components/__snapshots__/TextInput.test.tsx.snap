// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should render input field 1`] = `ReactWrapper {}`;

exports[`<TextInput /> should render input field with error 1`] = `ReactWrapper {}`;

exports[`<TextInput /> should render input field without error 1`] = `ReactWrapper {}`;
